zrml/neo-swaps/src/migration.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright 2023 Forecasting Technologies LTD.
+//
+// This file is part of Zeitgeist.
+//
+// Zeitgeist is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at
+// your option) any later version.
+//
+// Zeitgeist is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{
+    types::{LiquidityTree, Pool, SoloLp},
+    Config, Pallet,
+};
+use frame_support::{
+    dispatch::Weight,
+    log,
+    pallet_prelude::PhantomData,
+    traits::{Get, OnRuntimeUpgrade, StorageVersion},
+};
+use sp_runtime::traits::Saturating;
+
+cfg_if::cfg_if! {
+    if #[cfg(feature = "try-runtime")] {
+        use crate::{types::Node, MarketIdOf, Pools};
+        use alloc::{collections::BTreeMap, format, vec, vec::Vec};
+        use frame_support::{migration::storage_key_iter, pallet_prelude::Twox64Concat};
+        use parity_scale_codec::{Decode, Encode};
+        use sp_runtime::traits::Zero;
+    }
+}
+
+cfg_if::cfg_if! {
+    if #[cfg(any(feature = "try-runtime", test))] {
+        const NEO_SWAPS: &[u8] = b"NeoSwaps";
+        const POOLS: &[u8] = b"Pools";
+    }
+}
+
+const NEO_SWAPS_REQUIRED_STORAGE_VERSION: u16 = 0;
+const NEO_SWAPS_NEXT_STORAGE_VERSION: u16 = NEO_SWAPS_REQUIRED_STORAGE_VERSION + 1;
+
+type OldPoolOf<T> = Pool<T, SoloLp<T>>;
+
+pub struct MigrateToLiquidityTree<T>(PhantomData<T>);
+
+impl<T: Config + zrml_market_commons::Config> OnRuntimeUpgrade for MigrateToLiquidityTree<T> {
+    fn on_runtime_upgrade() -> Weight {
+        let mut total_weight = T::DbWeight::get().reads(1);
+        let market_commons_version = StorageVersion::get::<Pallet<T>>();
+        if market_commons_version != NEO_SWAPS_REQUIRED_STORAGE_VERSION {
+            log::info!(
+                "MigrateToLiquidityTree: market-commons version is {:?}, but {:?} is required",
+                market_commons_version,
+                NEO_SWAPS_REQUIRED_STORAGE_VERSION,
+            );
+            return total_weight;
+        }
+        log::info!("MigrateToLiquidityTree: Starting...");
+        let mut translated = 0u64;
+        crate::Pools::<T>::translate::<OldPoolOf<T>, _>(|_, pool| {
+            let solo = pool.liquidity_shares_manager;
+            // This should never fail; if it does, then we just delete the entry.
+            let mut liquidity_tree =
+                LiquidityTree::new(solo.owner.clone(), solo.total_shares).ok()?;
+            liquidity_tree.nodes.get_mut(0)?.fees = solo.fees; // Can't fail.
```

Comment:
We should at least log if one of those fails. We should also use `try-runtime` to ensure this case won't happen by using `pre_upgrade`.

Reply:
Which one should I log? They both 100% can't fail, so I'm having a hard time choosing.

Reply:
I get your point that deleting a pool would really suck. Just to be clear: If the first call fails, this means that there's a storage overflow due to someone misconfiguring the tree's depth. If the second call fails, this means that despit `new` pushing a root node onto the tree, that node is suddently gone in the next line. Seems fine to me. We make way more assumptions in other parts of the code.

Looking at this from another perspective: If this happens on a live chain, we would catch it early using try-runtime, because try-runtime verifies that all pools are properly migrated (in particular, not deleted). So the only risk is that between the last try-runtime run and actualy deployment, some user pushes a mystery pool which someone manages to make either of the calls above fail. But these calls have nothing to do with the content of the old pool. They're `LiquidityTree` calls. So I say we're fine here.